File: Android/sdk-weather-extension-android/weatherexample/src/main/java/com/adobe/marketing/mobile/weatherexample/WeatherExtensionInternal.java
/*
 ADOBE CONFIDENTIAL

 Copyright 2019 Adobe
 All Rights Reserved.

 NOTICE: Adobe permits you to use, modify, and distribute this file in
 accordance with the terms of the Adobe license agreement accompanying
 it. If you have received this file from a source other than Adobe,
 then your use, modification, or distribution of it requires the prior
 written permission of Adobe.
 */

package com.adobe.marketing.mobile.weatherexample;

import android.util.Log;

import com.adobe.marketing.mobile.AdobeCallback;
import com.adobe.marketing.mobile.Event;
import com.adobe.marketing.mobile.Extension;
import com.adobe.marketing.mobile.ExtensionApi;
import com.adobe.marketing.mobile.ExtensionError;
import com.adobe.marketing.mobile.ExtensionErrorCallback;
import com.adobe.marketing.mobile.ExtensionUnexpectedError;
import com.adobe.marketing.mobile.MobileCore;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

class WeatherExtensionInternal extends Extension {

    private static final String IDENTITY_MID_KEY = "mid";
    private static final String RULES_TRIGGERED_CONSEQUENCE_KEY = "triggeredconsequence";
    private static final String RULES_DETAIL_KEY = "detail";
    private static final String WEATHER_API_KEY = "weather.apiKey";
    private static final String WEATHER_ZIP_KEY = "zip";
    private static final String WEATHER_KEY = "weather";

    private ConcurrentLinkedQueue<Event> eventQueue;
    private WeatherExtensionApiHandler apiHandler;
    private ExecutorService executorService;
    private final Object executorMutex = new Object();
    private String mid;

    protected WeatherExtensionInternal(final ExtensionApi extensionApi) {
        super(extensionApi);

        // register a listener for shared state changes
        extensionApi.registerEventListener(
                WeatherExtensionConstants.EVENT_TYPE_ADOBE_HUB,
                WeatherExtensionConstants.EVENT_SOURCE_ADOBE_SHARED_STATE,
                WeatherExtensionListener.class, null);

        // register a listener for RulesEngine response events
        extensionApi.registerEventListener(
                WeatherExtensionConstants.EVENT_TYPE_ADOBE_RULES_ENGINE,
                WeatherExtensionConstants.EVENT_SOURCE_ADOBE_RESPONSE_CONTENT,
                WeatherExtensionListener.class, null);

        // register a listener for WeatherExtension request events
        extensionApi.registerEventListener(
                WeatherExtensionConstants.EVENT_TYPE_WEATHER_EXTENSION,
                WeatherExtensionConstants.EVENT_SOURCE_WEATHER_REQUEST_CONTENT,
                WeatherExtensionListener.class, null);

        // initialize the events queue
        eventQueue = new ConcurrentLinkedQueue<>();
        apiHandler = new WeatherExtensionApiHandler();
    }

    @Override
    protected String getName() {
        return "com.acpExample.weatherExtension";
    }

    @Override
    protected String getVersion() {
        return "1.0.0";
    }

    @Override
    protected void onUnregistered() {
        super.onUnregistered();

        // the extension was unregistered
        // if the shared states are not used in the next registration they can be cleared in this method
        getApi().clearSharedEventStates(null);
        mid = null;
    }

    @Override
    protected void onUnexpectedError(final ExtensionUnexpectedError unexpectedError) {
        super.onUnexpectedError(unexpectedError);
    }

    protected void queueEvent(final Event event) {
        if (event == null) {
            return;
        }

        eventQueue.add(event);
    }

    protected void processEvents() {
        while (!eventQueue.isEmpty()) {
            Event eventToProcess = eventQueue.peek();

            ExtensionErrorCallback<ExtensionError> extensionErrorCallback = new ExtensionErrorCallback<ExtensionError>() {
                @Override
                public void error(final ExtensionError extensionError) {
                    Log.e(getName(), String.format("Could not process event, an error occurred while retrieving configuration shared state: %s", extensionError.getErrorName()));
                }
            };
            Map<String, Object> configSharedState = getApi().getSharedEventState(WeatherExtensionConstants.SharedState.CONFIGURATION, eventToProcess, extensionErrorCallback);

            // NOTE: configuration is mandatory processing the event, so if shared state is null (pending) stop processing events
            if (configSharedState == null) {
                Log.d(getName(), "Could not process event, configuration shared state is pending");
                return;
            }

            String apiKey = configSharedState.isEmpty() ? "" : (String)configSharedState.get(WEATHER_API_KEY);

            // Identity is not a mandatory dependency for this event, just retrieve mid
            extractMidFromIdentitySharedState(eventToProcess);

            if (WeatherExtensionConstants.EVENT_TYPE_WEATHER_EXTENSION.equalsIgnoreCase(eventToProcess.getType())) {
                // handle the get weather event
                handleGetWeatherEvent(eventToProcess, apiKey);
                Log.d(getName(), "GetWeather event successfully processed");
            } else if (WeatherExtensionConstants.EVENT_TYPE_ADOBE_RULES_ENGINE.equalsIgnoreCase(eventToProcess.getType())) {
                // handle the rules consequence
                handleRulesConsequence(eventToProcess, apiKey);
                Log.d(getName(), "Rules consequence event successfully processed");
            }

            // event processed, remove it from the queue
            eventQueue.poll();
        }
    }

    private void handleGetWeatherEvent(final Event event, final String apiKey) {
        Log.d(getName(), "Handling Get Weather event");
        String zip = (String) event.getEventData().get(WEATHER_ZIP_KEY);

        apiHandler.getWeatherForZip(zip, apiKey, new AdobeCallback<Map<String, Object>>(){

            @Override
            public void call(final Map<String, Object> content) {
                // create the response event
                Log.d(getName(),String.format("Weather object returned: %s", content));
                Map<String, Object> eventData = new HashMap<>();
                eventData.put(WEATHER_KEY, content);
                Event weatherLoadedEvent = new Event.Builder("Weather Loaded",
                        WeatherExtensionConstants.EVENT_TYPE_WEATHER_EXTENSION,
                        WeatherExtensionConstants.EVENT_SOURCE_WEATHER_RESPONSE_CONTENT)
                    .setEventData(eventData).build();

                if (weatherLoadedEvent == null) {
                    Log.w(getName(), "An error occurred constructing event");
                }

                // update weather extension's shared state
                getApi().setSharedEventState(content, event, null);

                // dispatch the response for the public API
                ExtensionErrorCallback<ExtensionError> errorCallback = new ExtensionErrorCallback<ExtensionError>() {
                    @Override
                    public void error(final ExtensionError e) {
                        Log.w(getName(), String.format("An error occurred dispatching the weather loaded event: %s", e.getErrorName()));
                    }
                };
                MobileCore.dispatchResponseEvent(weatherLoadedEvent, event, errorCallback);
            }
        });
    }

    private void handleRulesConsequence(final Event event, final String apiKey) {
        Log.d(getName(), "Handling Triggered Consequence event");
        Map<String, Object> eventData = event.getEventData();

        Map<String, Object> triggeredConsequence = (HashMap<String, Object>) eventData.get(RULES_TRIGGERED_CONSEQUENCE_KEY);
        if (triggeredConsequence == null || triggeredConsequence.isEmpty()) {
            return;
        }

        Map<String, Object> detail = (Map<String, Object>)triggeredConsequence.get(RULES_DETAIL_KEY);
        if (detail == null || detail.isEmpty()) {
            return;
        }

        String zip = (String) detail.get(WEATHER_ZIP_KEY);
        if (zip == null) {
            Log.d(getName(), "Not a zip request consequence");
            return;
        }

        apiHandler.getWeatherForZip(zip, apiKey, new AdobeCallback<Map<String, Object>>() {

            @Override
            public void call(final Map<String, Object> content) {
                Log.d(getName(),String.format("The weather has been updated for this user: %s, %s", content.get("temp"), content.get("conditions")));
                // update weather extension's shared state

                ExtensionErrorCallback<ExtensionError> setSharedStateError = new ExtensionErrorCallback<ExtensionError>() {
                    @Override
                    public void error(ExtensionError extensionError) {
                        Log.e(getName(), String.format("An error occurred while setting the shared state %s, error %s", content, extensionError.getErrorName()));
                    }
                };

                getApi().setSharedEventState(content, event,  setSharedStateError);
            }
        });
    }

    private void extractMidFromIdentitySharedState(final Event event) {
        Map<String, Object> identitySharedState = getApi().getSharedEventState(WeatherExtensionConstants.SharedState.IDENTITY, event, null);

        if (identitySharedState == null) {
            Log.d(getName(), "Identity shared state is pending, returning nil");
            return;
        }

        this.mid = (String) identitySharedState.get(IDENTITY_MID_KEY);
    }

    protected final ExecutorService getExecutor() {
        synchronized (executorMutex) {
            if (executorService == null) {
                executorService = Executors.newSingleThreadExecutor();
            }

            return executorService;
        }
    }
}

File: iOS/ACPWeatherExample/ACPWeatherExample/WeatherExtensionInternal.m
/*
 ADOBE CONFIDENTIAL
 
 Copyright 2019 Adobe
 All Rights Reserved.
 
 NOTICE: Adobe permits you to use, modify, and distribute this file in
 accordance with the terms of the Adobe license agreement accompanying
 it. If you have received this file from a source other than Adobe,
 then your use, modification, or distribution of it requires the prior
 written permission of Adobe.
 */

#import "ACPCore.h"
#import "ACPExtensionApi.h"
#import "ACPExtensionEvent.h"
#import "WeatherExtensionInternal.h"
#import "WeatherExtensionListener.h"
#import "WeatherExtensionApiHandler.h"
#import "WeatherExtensionDataObject.h"
#import "Queue.h"

@interface WeatherExtensionInternal()
@property (nonatomic, strong) WeatherExtensionApiHandler *apiHandler;
@property (nonatomic, strong) Queue *eventQueue;
@property (nonatomic, strong) NSString *mid;
@end

@implementation WeatherExtensionInternal

static NSString* LOG_TAG = @"WeatherExtensionInternal";
static NSString* ACP_STATE_OWNER = @"stateowner";
static NSString* ACP_CONFIGURATION_SHARED_STATE = @"com.adobe.module.configuration";
static NSString* ACP_IDENTITY_SHARED_STATE = @"com.adobe.module.identity";
static NSString* ACP_IDENTITY_MID_KEY = @"mid";
static NSString* ACP_RULES_TRIGGERED_CONSEQUENCE_KEY = @"triggeredconsequence";
static NSString* ACP_RULES_DETAIL_KEY = @"detail";
static NSString* WEATHER_API_KEY = @"weather.apiKey";
static NSString* WEATHER_ZIP_KEY = @"zip";
static NSString* WEATHER_KEY = @"weather";

#pragma mark - ACPExtension methods
- (nullable NSString *) name {
    return @"com.acpExample.weatherExtension";
}

- (NSString *) version {
    return @"1.0.0";
}

- (instancetype) init {
    if (self = [super init]) {
        NSError *error = nil;
        
        // register a listener for shared state changes
        if ([self.api registerListener:[WeatherExtensionListener class]
                             eventType:@"com.adobe.eventType.hub"
                           eventSource:@"com.adobe.eventSource.sharedState"
                                 error:&error]) {
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:@"WeatherExtensionListener successfully registered for Event Hub Shared State events"];
        } else {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"There was an error registering WeatherExtensionListener for Event Hub Shared State events: %@", error.localizedDescription ?: @"unknown"]];
        }
        
        // register a listener for RulesEngine response events
        error = nil;
        if ([self.api registerListener:[WeatherExtensionListener class]
                             eventType:@"com.adobe.eventType.rulesEngine"
                           eventSource:@"com.adobe.eventSource.responseContent"
                                 error:&error]) {
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:@"WeatherExtensionListener successfully registered for Rules Engine Response Content events"];
        } else {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"There was an error registering WeatherExtensionListener for Rules Engine Response Content events: %@", error.localizedDescription ?: @"unknown"]];
        }
        
        // register a listener for WeatherExtension request events
        error = nil;
        if ([self.api registerListener:[WeatherExtensionListener class]
                             eventType:@"com.acpExample.eventType.weatherExtension"
                           eventSource:@"com.acpExample.eventSource.requestContent"
                                 error:&error]) {
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:@"WeatherExtensionListener successfully registered for WeatherExtension Request Content events"];
        } else {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"There was an error registering WeatherExtensionListener for WeatherExtension Request Content events: %@", error.localizedDescription ?: @"unknown"]];
        }
        
        // create our WeatherExtensionApiHandler
        self.apiHandler = [[WeatherExtensionApiHandler alloc] init];
        
        // initialize the events queue
        self.eventQueue = [[Queue alloc] init];
    }

    return self;
}

- (void) onUnregister {
    [super onUnregister];
    
    // the extension was unregistered
    // if the shared states are not used in the next registration they can be cleared in this method
    [[self api] clearSharedEventStates:nil];
    
    self.apiHandler.apiKey = @"";
    self.mid = nil;
}

- (void) unexpectedError:(NSError *)error {
    [super unexpectedError:error];
}

#pragma mark - WeatherExtension Internal methods
- (void) handleGetWeatherEvent:(ACPExtensionEvent *)event {
    [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Handling Get Weather event", [self name]]];
    NSNumber *zip = event.eventData[WEATHER_ZIP_KEY];
    [self.apiHandler getWeatherForZip:[zip integerValue] callback:^(NSDictionary* _Nullable content) {
        // create the response event
        [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Weather object returned: %@", [self name], content]];
        
        NSError *eventError = nil;
        NSDictionary* eventData = @{WEATHER_KEY : content};
        ACPExtensionEvent *weatherLoadedEvent = [ACPExtensionEvent extensionEventWithName:@"Weather Loaded"
                                                                                     type:@"com.acpExample.eventType.weatherExtension"
                                                                                   source:@"com.acpExample.eventSource.responseContent"
                                                                                     data:eventData
                                                                                    error:&eventError];
        if (!weatherLoadedEvent) {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - An error occurred constructing event '%@': %@", [self name], weatherLoadedEvent.eventName, eventError.localizedDescription ?: @"unknown"]];
        }
        
        // update weather extension's shared state
        NSError *setSharedStateError = nil;
        if (![[self api] setSharedEventState:content event:event error:&setSharedStateError] && setSharedStateError != nil) {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"An error occurred while setting the shared state %@, error code %ld", content, [setSharedStateError code]]];
        }
        
        // dispatch the response for the public API
        NSError *dispatchError = nil;
        if ([ACPCore dispatchResponseEvent:weatherLoadedEvent requestEvent:event error:&dispatchError]) {
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"WeatherExtension dispatched an event '%@'", weatherLoadedEvent.eventName]];
        } else {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - An error occurred dispatching event '%@': %@", [self name], weatherLoadedEvent.eventName, dispatchError.localizedDescription ?: @"unknown"]];
        }
    }];
}

- (void) handleRulesConsequence:(ACPExtensionEvent *)event {
    [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Handling Triggered Consequence event", [self name]]];
    NSDictionary *eventData = event.eventData;
    if (!eventData) {
        return;
    }
    
    NSDictionary *triggeredConsequence = eventData[ACP_RULES_TRIGGERED_CONSEQUENCE_KEY];
    if (!triggeredConsequence || [triggeredConsequence count] == 0) {
        return;
    }
    
    NSDictionary *detail = triggeredConsequence[ACP_RULES_DETAIL_KEY];
    if (!detail || [detail count] == 0) {
        return;
    }

    NSString *zip = detail[WEATHER_ZIP_KEY];
    if (!zip) {
        [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Not a zip request consequence", [self name]]];
        return;
    }
    
    [self.apiHandler getWeatherForZip:[zip integerValue] callback:^(NSDictionary* _Nullable content) {
        [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - The weather has been updated for this user: %@, %@", self.name, content[@"temp"], content[@"conditions"]]];
        // update weather extension's shared state
        NSError *setSharedStateError = nil;
        if (![[self api] setSharedEventState:content event:event error:&setSharedStateError] && setSharedStateError != nil) {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"An error occurred while setting the shared state %@, error code %ld", content, [setSharedStateError code]]];
        }
    }];
}

- (void) queueEvent: (ACPExtensionEvent*) event {
    if (!event) {
        return;
    }
    
    [self.eventQueue add:event];
}

- (void) processEvents {
    while ([self.eventQueue hasNext]) {
        ACPExtensionEvent* eventToProcess = [self.eventQueue peek];
        
        NSError *error = nil;
        NSDictionary *configSharedState = [self.api getSharedEventState:ACP_CONFIGURATION_SHARED_STATE event:eventToProcess error:&error];
        
        // NOTE: configuration is mandatory processing the event, so if shared state is null stop processing events
        if (!configSharedState) {
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Could not process event, configuration shared state is pending", [self name]]];
            return;
        }
        
        if (error != nil) {
            [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Could not process event, an error occured while retrieving configuration shared state %ld", [self name], [error code]]];
            return;
        }
        
        self.apiHandler.apiKey = configSharedState.count ? configSharedState[WEATHER_API_KEY] : @"";
        
        // Identity is not a mandatory dependency for this event, just retrive mid
        [self extractMidFromIdentitySharedState:eventToProcess];
        
        if ([eventToProcess.eventType isEqualToString:@"com.acpExample.eventType.weatherExtension"]) {
            // handle the get weather event
            [self handleGetWeatherEvent:eventToProcess];
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - GetWeather event successfully processed", [self name]]];
        } else if ([eventToProcess.eventType isEqualToString:@"com.adobe.eventType.rulesEngine"]) {
            // handle the rules consequence
            [self handleRulesConsequence:eventToProcess];
            [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Rules consequence event successfully processed", [self name]]];
        }
        [self.eventQueue poll];
    }
}

- (void) extractMidFromIdentitySharedState:(ACPExtensionEvent *) event {
    NSError *error = nil;
    NSDictionary *identitySharedState = [self.api getSharedEventState:ACP_IDENTITY_SHARED_STATE event:event error:&error];
    
    if (!identitySharedState) {
        [ACPCore log:ACPMobileLogLevelDebug tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - Identity shared state is pending, returning nil", [self name]]];
        return;
    }
    
    if (error != nil) {
        [ACPCore log:ACPMobileLogLevelError tag:LOG_TAG message:[NSString stringWithFormat:@"%@ - An error occured while retrieving identity shared state %ld", [self name], [error code]]];
        return;
    }
    
    if (identitySharedState.count) {
        self.mid = identitySharedState[ACP_IDENTITY_MID_KEY];
    }
}

@end
